spatialos-sdk/src/worker/entity.rs

```diff
@@ -36,36 +45,151 @@ impl Entity {
             user_handle: data_ptr as *mut _,
         };
 
+        self.components.insert(C::ID, raw_data);
+
+        Ok(())
+    }
+
+    pub(crate) fn add_raw(&mut self, component: &Worker_ComponentData) -> Result<(), String> {
+        let id = component.component_id;
+
+        self.pre_add_check(id)?;
+
+        // Call copy on the component data. We don't own this Worker_ComponentData.
+        let vtable = self.database.get_vtable(id).unwrap();
+        let copy_data_func = vtable
+            .component_data_copy
+            .unwrap_or_else(|| panic!("No component_data_free method defined for {}", id));
+        unsafe { copy_data_func(id, ptr::null_mut(), component.user_handle) };
+
         self.components.insert(
-            C::ID,
-            ComponentData {
-                raw_data,
-                drop_fn: drop_raw::<C>,
+            id,
+            Worker_ComponentData {
+                reserved: ptr::null_mut(),
+                component_id: id,
+                schema_type: ptr::null_mut(),
+                user_handle: component.user_handle,
             },
         );
+
+        Ok(())
     }
 
     pub fn get<C: Component>(&self) -> Option<&C> {
         self.components
             .get(&C::ID)
-            .map(|data| unsafe { &*(data.raw_data.user_handle as *const _) })
+            .map(|data| unsafe { &*(data.user_handle as *const _) })
+    }
+
+    pub(crate) fn raw_component_data(&self) -> RawEntity {
+        RawEntity::new(self.components.values())
+    }
+
+    fn pre_add_check(&self, id: ComponentId) -> Result<(), String> {
+        if self.components.contains_key(&id) {
+            return Err(format!(
+                "Duplicate component with ID {} added to `Entity`.",
+                id
+            ));
+        }
+
+        if !self.database.has_vtable(id) {
+            return Err(format!(
+                "Could not find a vtable implementation for component {}",
```

Comment:
Does it make sense to return an error if the vtable is missing? With the way the SDK is setup currently, this could only be caused by a misconfiguration by the user, i.e. they didn't register a component. It might be more appropriate to panic in this case.

Reply:
Yeah a panic may be more appropriate - also reduces the need to `unwrap()` empty returns